src/reason_parser.mly

```diff
@@ -3205,6 +3212,13 @@ mark_position_pat
     * semantics (they are not first class).
     */
   | as_loc(constr_longident) pattern_constructor_argument
+    /* the first case is `| Foo(_)` and doesn't need explicit_arity attached. Actually, something like `| Foo(1)` doesn't either, but we 
+      keep explicit_arity on the latter anyways because why not. But for `| Foo(_)` in particular, it's convenient to have explicit_arity 
```

Comment:
It would be nice if we could avoid the allocation of an extra tuple. ReasonReact might benefit from this. e.g. reducers with variants with one arg (which is pretty common in my code, not sure if it's the case with other's too)

I think it should be pretty easy if you modify `simple_pattern_list_to_tuple` in
```
  | as_loc(constr_longident) pattern_constructor_argument
    { match is_pattern_list_single_any $2 with
      | Some singleAnyPat ->
        mkpat (Ppat_construct($1, Some singleAnyPat))
      | None ->
        let loc = mklocation $symbolstartpos $endpos in
        let argPattern = simple_pattern_list_to_tuple ~loc $2 in
        mkExplicitArityTuplePat (Ppat_construct($1, Some argPattern))
    }
```
to not be called if there's only one arg.

Reply:
I thought the allocation was compiled away at a later stage and that's why we didn't bother before. @jordwalke

Reply:
![4248e709-9347-432f-8cea-f90cb0850d3c](https://user-images.githubusercontent.com/1909539/31689328-f266b67e-b343-11e7-8100-6219a804141a.jpeg)

Reply:
Yeah, bs seems to remove the boxing, so that makes me pretty confident that it's removed in ocamlopt too. We can also dump the later compiler stages to ensure that it is the case.

Reply:
Should have checked Bucklescript's output 😅